### PR TITLE
GIG-1955 avoid unnecessary rebuilds

### PR DESCRIPTION
**Pull request by billvb**

### CODE REVIEW

File: lmcommon/container/core.py [hunk 1]
```diff
@@ -92,6 +143,13 @@ def build_docker_image(root_dir: str, override_image_tag: Optional[str], nocache
                                                                owner=lb.owner['username'],
                                                                username=username)
 
+    reuse_image_id = _get_cached_image(env_dir, image_name)
+    if reuse_image_id:
+        logger.warning(f"Reusing Docker image for {str(lb)}")
```
Comment (dkleissa) [inline lmcommon/container/core.py]: Probably doesn't nee to be a warning

File: lmcommon/container/tests/test_container_ops.py [hunk 2]
```diff
@@ -96,6 +97,8 @@ def test_old_dockerfile_removed_when_new_build_fails(self, build_lb_image_for_ju
             dockerfile.write('\n'.join(olines))
             dockerfile.write('\nRUN /bin/false')
 
+        [_ for _ in remove_image_cache_data()]
```
Comment (dkleissa) [inline lmcommon/container/tests/test_container_ops.py]: this is a weird way to use a fixture? Why not just pass it in and let the cleanup run like normal?

File: lmcommon/container/tests/test_container_ops.py [hunk 3]
```diff
@@ -96,6 +104,9 @@ def test_old_dockerfile_removed_when_new_build_fails(self, build_lb_image_for_ju
             dockerfile.write('\n'.join(olines))
             dockerfile.write('\nRUN /bin/false')
 
+        # We need to remove cache data otherwise the following tests won't work
+        remove_image_cache_data()
```
Comment (dkleissa) [inline lmcommon/container/tests/test_container_ops.py]: Should there be a test making sure the hash file is saved and that it changes if you change the env and rebuild and it doesn't if you don't change the env and rebuild?

Comment (billvb) [reply lmcommon/container/tests/test_container_ops.py]: So i've been thinking about for a while this morning now. I can't find a good place to squeeze it in. We do have it in tester-tests though, FWIW

File: lmcommon/container/core.py [hunk 4]
```diff
@@ -117,9 +178,11 @@ def build_docker_image(root_dir: str, override_image_tag: Optional[str], nocache
                 # There is no other (simple) way to grab the image ID
                 image_id = stream.split(' ')[-1]
     except docker.errors.BuildError as e:
+        _remove_docker_image(image_name)
```
Comment (dkleissa) [inline lmcommon/container/core.py]: possible downside of removing here is that you can't spin up the intermediate image and try to manually fix things if you're a power user. Probably fine for now...and won't be possible when running in the cloud anyway.

**Pull request merged**